### ISSUE TITLE
Question weighted gradients

### ISSUE THREAD

**Issue by mickvanhulst** [ManiadisG/Customising-Tensorflow-2.2-Models#1]
Hi,

Thanks for sharing this Github, it helped me with defining some custom training functions. I am still wondering, however, if it is possible to pass additional arguments that could be used to e.g. apply some weights to the gradient? In such a scenario we would have to e.g. pass additional arguments via `fit` to `train_step`?

Thanks in advance!

**Comment by ManiadisG** [comment]
Hello.

I will go through all three cases: immutable argument, mutable argument, input dependent argument.

1. For immutable arguments you can define them as a model attribute and use them in the `train_step function`. Fairly straightforward.

2. If you want to change the arguments at times (e.g. after some epoch via a callback) you must define them as a `tf.Variable` type model attribute with `trainable=False` and modify them `via tf.assign`. That is required for the graph to register the change.

3. If your arguments are input dependent (e.g. sample-specific weights) you would need to define an appropriate `tf.dataset`. For example:
`dataset = tf.data.Dataset.from_tensor_slices((input, ground_truth, arguments))`
In that case the first line of train_step would be:
`x, y, args = data`
Then you run fit as usual with that dataset. You can have more control over the dataset's output using `dataset.map` and folllowing the guide: https://www.tensorflow.org/guide/data
The reason why you would need to do that via a `tf.dataset` is that the `train_step` only takes in one input argument straight from the `dataset`. Changing that is not allowed.

I hope this is helpful. Please let me know if I was unclear someplace or if I didn't cover your question.

**Comment by mickvanhulst** [comment]
Awesome, thanks for your answer!
